docs/en/options.md
# About the configuration file

The directory where the configuration files are placed is created as `.twnyan` directly under the home directory.

```
.twnyan
├── .cred.yaml
├── color.yaml
└── option.yaml
```

## .cred.yaml

A file containing account credentials.

## option.yaml

Optional settings file.

| Name       | Description                         |
| ---------- | ----------------------------------- |
| ConfigDir  | Path of the configuration directory |
| Counts     | Default number of acquisitions      |
| DateFormat | Date Format                         |
| TimeFormat | Time Format                         |

- Date and time formats are in the same format as the format string in the [time package](https://golang.org/pkg/time/#pkg-constants)

## color.yaml

Color setting file.

| Name         | Description                    |
| ------------ | ------------------------------ |
| Accent1      | Accent Color 1                 |
| Accent2      | Accent Color 2                 |
| Accent3      | Accent Color 3                 |
| Error        | Error Message Background Color |
| BoxForground | Text Color When Reversed       |
| Separator    | Separator                      |
| UserName     | User Name                      |
| ScreenName   | Screen Name                    |
| Reply        | Reply                          |
| Hashtag      | Hashtag                        |
| Favorite     | Like                           |
| Retweet      | Retweet                        |
| Verified     | Verified Account               |
| Protected    | Private Account                |
| Following    | Following                      |
| FollowedBy   | Followed By                    |
| Block        | Block                          |
| Mute         | Mute                           |


docs/ja/options.md
# 設定ファイルについて

設定ファイルが置かれるディレクトリは、ホームディレクトリ直下に `.twnyan` として作成されます。

```
.twnyan
├── .cred.yaml
├── color.yaml
└── option.yaml
```

## .cred.yaml

アカウントの認証情報が含まれたファイルです。

## option.yaml

オプション設定のファイルです。

| 名前       | 説明                   |
| ---------- | ---------------------- |
| ConfigDir  | 設定ディレクトリのパス |
| Counts     | デフォルトの取得件数   |
| DateFormat | 日付のフォーマット     |
| TimeFormat | 時刻のフォーマット     |

- 日付、時刻のフォーマットは[time パッケージのフォーマット文字列](https://golang.org/pkg/time/#pkg-constants)と同じ書式です

## color.yaml

色設定のファイルです。

| 名前         | 説明                   |
| ------------ | ---------------------- |
| Accent1      | アクセント 1           |
| Accent2      | アクセント 2           |
| Accent3      | アクセント 3           |
| Error        | エラーメッセージ背景色 |
| BoxForground | 反転時の文字色         |
| Separator    | セパレータ             |
| UserName     | ユーザ名               |
| ScreenName   | スクリーンネーム       |
| Reply        | リプライ表示           |
| Hashtag      | ハッシュタグ           |
| Favorite     | いいね表示             |
| Retweet      | リツイート表示         |
| Verified     | 認証済みアカウント     |
| Protected    | 鍵アカウント           |
| Following    | フォロー中表示         |
| FollowedBy   | 被フォロー表示         |
| Block        | ブロック表示           |
| Mute         | ミュート表示           |


CHANGELOG.md
# Change Log

## [Unreleased]

## [1.8.2] - 2022-05-10

### Fixed

- 一定の条件を満たすハッシュタグ付きツイートを表示するとエラーが発生する

## [1.8.1] - 2022-05-04

### Changed

- `account list` の表示形式を変更

## [1.8.0] - 2022-05-04

### Security

- 依存関係を更新

### Added

- マルチアカウントに対応 [#15](https://github.com/arrow2nd/twnyan/issues/15)
- `tweet` コマンドをパイプ入力に対応 [#16](https://github.com/arrow2nd/twnyan/issues/16)

### Changed

- ヘルプ文を変更
- ドキュメントを整備

### Fixed

- 一部ハッシュタグがハイライトされない
- ツイートが無いユーザのタイムラインを表示するとエラーになる

## [1.7.2] - 2022-03-29

### Security

- 依存関係を更新

## [1.7.1] - 2021-12-14

### Fixed

- ツイート番号を指定してのユーザ操作ができない

## [1.7.0] - 2021-12-12

### Added

- いいねと RT を同時に行う likert コマンドを追加

## [1.6.0] - 2021-11-14

### Changed

- エラーメッセージの出力先を変更

## [1.5.2] - 2021-11-08

### Changed

- 自動リリースの環境を変更

## [1.5.1] - 2021-11-08

### Added

- Homebrew でのインストールに対応

### Fixed

- ハッシュタグのハイライトを修正

## [1.5.0] - 2021-08-12

### Changed

- multi コマンドの仕様を変更

## [1.4.6] - 2021-07-01

### Fixed

- 今日の投稿されたツイートに日付が表示される

## [1.4.5] - 2021-04-27

### Fixed

- アンダースコア・一部の漢字を含むハッシュタグが正しくハイライトされない

## [1.4.4] - 2021-04-19

### Fixed

- ハイライトされないハッシュタグがある

## [1.4.3] - 2021-04-13

### Fixed

- コマンドが見つからなかったときの挙動を修正

## [1.4.2] - 2021-04-13

### Fixed

- stream コマンド実行中にレート制限に引っかかった時の挙動を修正

## [1.4.1] - 2021-04-13

### Fixed

- デバッグログの消し忘れを修正

### Changed

- アカウント種類の表示形式を変更

## [1.4.0] - 2021-04-12

### Added

- UserStream を再現する stream コマンドを追加

## [1.3.0] - 2021-04-03

### Added

- 複数行入力をキャンセルする機能を追加

## [1.2.5] - 2021-03-22

### Fixed

- ハイライトされないハッシュタグがある

## [1.2.4] - 2021-03-13

### Fixed

- Go 1.16.x で go install が失敗する

## [1.2.3] - 2021-02-12

### Fixed

- ハッシュタグ等が途中で改行されるとハイライトされない
- アップロード処理を修正

## [1.2.2] - 2021-02-04

### Added

- quote コマンドを追加
- 複数行のツイートに画像添付機能を追加
- リプライ、引用 RT も複数行に対応

### Fixed

- ヘルプ文を修正

## [1.2.1] - 2021-02-03

### Fixed

- Windows 環境で起動できない
- リツイートの取り消しに失敗する
- API のエラーメッセージがパースできない

### Changed

- API 制限時にエラーメッセージを表示するよう変更

## [1.2.0] - 2021-02-03

### Added

- 複数行のツイートに対応

### Removed

- Export コマンドを削除
- Config コマンドを削除

### Changed

- 表示のデザインを変更
- 設定ファイルのパス、形式を変更

## [1.1.1] - 2020-12-30

### Added

- 設定ファイル操作時の確認を追加

### Changed

- デフォルト色を変更
- 一部デザインを変更

## [1.1.0] - 2020-12-18

### Added

- json/yaml 形式でのファイル出力機能

### Removed

- 形式を指定して取得したデータを標準出力する機能

### Changed

- コマンドのヘルプ文
- システムメッセージの表示形式

## [1.0.2] - 2020-12-17

### Fixed

- メンションのハイライトが正しくない

## [1.0.1] - 2020-12-17

### Fixed

- ハッシュタグのハイライトが正しくない

## [1.0.0] - 2020-12-03

### Fixed

- テキスト色が反映されない

### Removed

- .twnyan.yaml 内の色の設定からテキストの項目を削除

## [0.0.4] - 2020-12-01

### Fixed

- ハッシュタグ・メンションのハイライトが正しくない

## [0.0.3] - 2020-11-20

### Added

- 形式を指定して取得したデータを標準出力する機能

### Fixed

- ハッシュタグのハイライトが正しくない

## [0.0.2] - 2020-11-19

### Fixed

- 日本語環境での絵文字入力時の挙動がおかしい
- `go get`でインストールした際、正しい依存パッケージが取得できない

## 0.0.1 - 2020-11-17

- リリースしました！

[unreleased]: https://github.com/arrow2nd/twnyan/compare/v1.8.2...HEAD
[1.8.2]: https://github.com/arrow2nd/twnyan/compare/v1.8.1...v1.8.2
[1.8.1]: https://github.com/arrow2nd/twnyan/compare/v1.8.0...v1.8.1
[1.8.0]: https://github.com/arrow2nd/twnyan/compare/v1.7.2...v1.8.0
[1.7.2]: https://github.com/arrow2nd/twnyan/compare/v1.7.1...v1.7.2
[1.7.1]: https://github.com/arrow2nd/twnyan/compare/v1.7.0...v1.7.1
[1.7.0]: https://github.com/arrow2nd/twnyan/compare/v1.6.0...v1.7.0
[1.6.0]: https://github.com/arrow2nd/twnyan/compare/v1.5.2...v1.6.0
[1.5.2]: https://github.com/arrow2nd/twnyan/compare/v1.5.1...v1.5.2
[1.5.1]: https://github.com/arrow2nd/twnyan/compare/v1.5.0...v1.5.1
[1.5.0]: https://github.com/arrow2nd/twnyan/compare/v1.4.6...v1.5.0
[1.4.6]: https://github.com/arrow2nd/twnyan/compare/v1.4.5...v1.4.6
[1.4.5]: https://github.com/arrow2nd/twnyan/compare/v1.4.4...v1.4.5
[1.4.4]: https://github.com/arrow2nd/twnyan/compare/v1.4.3...v1.4.4
[1.4.3]: https://github.com/arrow2nd/twnyan/compare/v1.4.2...v1.4.3
[1.4.2]: https://github.com/arrow2nd/twnyan/compare/v1.4.1...v1.4.2
[1.4.1]: https://github.com/arrow2nd/twnyan/compare/v1.4.0...v1.4.1
[1.4.0]: https://github.com/arrow2nd/twnyan/compare/v1.3.0...v1.4.0
[1.3.0]: https://github.com/arrow2nd/twnyan/compare/v1.2.5...v1.3.0
[1.2.5]: https://github.com/arrow2nd/twnyan/compare/v1.2.4...v1.2.5
[1.2.4]: https://github.com/arrow2nd/twnyan/compare/v1.2.3...v1.2.4
[1.2.3]: https://github.com/arrow2nd/twnyan/compare/v1.2.2...v1.2.3
[1.2.2]: https://github.com/arrow2nd/twnyan/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/arrow2nd/twnyan/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/arrow2nd/twnyan/compare/v1.1.1...v1.2.0
[1.1.1]: https://github.com/arrow2nd/twnyan/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/arrow2nd/twnyan/compare/v1.0.2...v1.1.0
[1.0.2]: https://github.com/arrow2nd/twnyan/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/arrow2nd/twnyan/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/arrow2nd/twnyan/compare/v0.0.4...v1.0.0
[0.0.4]: https://github.com/arrow2nd/twnyan/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/arrow2nd/twnyan/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/arrow2nd/twnyan/compare/v0.0.1...v0.0.2


config/config.go
package config

import "github.com/garyburd/go-oauth/oauth"

const (
	crdFile = ".cred.yaml"
	optFile = "option.yaml"
	colFile = "color.yaml"
)

type cred struct {
	// Main メインアカウント
	Main *oauth.Credentials `yaml:"Main,inline"`
	// Sub サブアカウント
	Sub map[string]*oauth.Credentials
}

type option struct {
	ConfigDir  string
	Counts     string
	DateFormat string
	TimeFormat string
}

type color struct {
	Accent1      string
	Accent2      string
	Accent3      string
	Error        string
	BoxForground string
	Separator    string
	UserName     string
	ScreenName   string
	Reply        string
	Hashtag      string
	Favorite     string
	Retweet      string
	Verified     string
	Protected    string
	Following    string
	FollowedBy   string
	Block        string
	Mute         string
}

// Config 設定管理
type Config struct {
	// Cred 認証情報
	Cred *cred
	// Option 設定情報
	Option *option
	// Color 色情報
	Color *color
}

// New 生成
func New() *Config {
	return &Config{
		Cred: &cred{
			Main: &oauth.Credentials{},
			Sub:  map[string]*oauth.Credentials{},
		},
		Option: &option{
			ConfigDir:  getConfigDir(),
			Counts:     "25",
			DateFormat: "2006/01/02",
			TimeFormat: "15:04:05",
		},
		Color: &color{
			Accent1:      "#e06c75",
			Accent2:      "#c678dd",
			Accent3:      "#56b6c2",
			Error:        "#e06c75",
			BoxForground: "#000000",
			Separator:    "#707070",
			UserName:     "#faf8f7",
			ScreenName:   "#9c9c9c",
			Reply:        "#56b6c2",
			Hashtag:      "#61afef",
			Favorite:     "#e887b9",
			Retweet:      "#98c379",
			Verified:     "#5685d1",
			Protected:    "#787878",
			Following:    "#1877c9",
			FollowedBy:   "#18a0c9",
			Block:        "#e06c75",
			Mute:         "#e5c07b",
		},
	}
}

// Save 保存
func (cfg *Config) Save() {
	cfg.saveYaml(crdFile, cfg.Cred)
	cfg.saveYaml(optFile, cfg.Option)
	cfg.saveYaml(colFile, cfg.Color)
}

// Load 読込
func (cfg *Config) Load() bool {
	if !cfg.configFileExists() {
		return false
	}

	cfg.loadYaml(crdFile, cfg.Cred)
	cfg.loadYaml(optFile, cfg.Option)
	cfg.loadYaml(colFile, cfg.Color)

	return true
}
